### ISSUE TITLE
Completion of error handling

### ISSUE THREAD

**Issue by elfring** [EvilPudding/candle#2]
Would you like to add more error handling for return values from functions like the following?
- [malloc](http://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html "Memory allocation") ⇒ [vector_new](https://github.com/EvilPudding/candle/blob/97e9162c891bb0c6c38e691fc71327383dc007e6/utils/vector.c#L29)
- [strdup](http://pubs.opengroup.org/onlinepubs/9699919799/functions/strdup.html "Duplicate a string") ⇒ [vertex_modifier_new](https://github.com/EvilPudding/candle/blob/ed743ab69c28b71b9020d1705f0e92dc36f858de/utils/shader.c#L137)

**Comment by EvilPudding** [comment]
Sure, good suggestion.

**Comment by elfring** [comment]
I suggest to avoid [ignorance of return values](https://cwe.mitre.org/data/definitions/252.html "Checking more return values") a bit more.
Would you like to detect every error situation as early as possible?

**Comment by EvilPudding** [comment]
Can you give me an example of such a case?

**Comment by elfring** [comment]
[Is the implementation of the function “shader_new_loader” another update candidate](https://github.com/EvilPudding/candle/blob/ed743ab69c28b71b9020d1705f0e92dc36f858de/utils/shader.c#L414 "Update candidate: shader_new_loader()")?

-----

How do you think about to improve [static source code analysis](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2C_C.2B.2B "Tools for code analysis") also for this software?

**Comment by EvilPudding** [comment]
I don't understand at all what you are saying, could you rephrase?

**Comment by elfring** [comment]
* Do you care which function calls can return a null pointer (or a zero value)?
* How would you like to handle possible error situations?

**Comment by EvilPudding** [comment]
It depends on the function. If it's a function which needs to return a non-zero value for the engine to proceed, then the program crashing is good because it triggers gdb so the error can be debugged, if it's a function that is expected to return zero, then there is nothing we should do.

**Comment by elfring** [comment]
I suggest to reconsider such a view so that [undefined software behaviour will be avoided](https://wiki.sei.cmu.edu/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers "Do not dereference null pointers.").

**Comment by EvilPudding** [comment]
Many functions expect to receive non-zero pointers, it's one of the advantages of C is that the language doesn't enforce it, so you can optimize those checks out.

Why would I want to remove undefined behaviour that would come up in case of bad usage of the function? If I call `strlen` with zero, I get a segmentation fault, and I should.

**Comment by elfring** [comment]
Should it be generally avoided to pass inappropriate parameters for specific function calls?

**Comment by EvilPudding** [comment]
Of course.

**Comment by elfring** [comment]
Will any checks be required for the handling of failure predicates from more function calls then?

**Comment by EvilPudding** [comment]
That is vague, I don't know, if you find anything specific, we can discuss it.

**Comment by elfring** [comment]
Are the links for three questionable source code places concrete enough already?

**Comment by EvilPudding** [comment]
The `malloc` link and `strdup` I understand, I don't understand what you meant about `shader_new_loader`

**Comment by elfring** [comment]
Would you like to take descriptions also for programming interfaces like [OpenGL](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/start.html "OpenGL® reference pages") better into account?

“…
glCreateProgram
…
This function returns 0 if an error occurs creating the program object.
…”

**Comment by EvilPudding** [comment]
That is being handled, `glerr` checks for an OpenGL error and quits the program if one is found.

**Comment by elfring** [comment]
Will it be better to check the return value also from such a function call directly instead of using [the function “_check_gl_error”](https://github.com/EvilPudding/candle/blob/4e40b4022d9284e0e5c9b5764da4b1272a984f25/utils/glutil.c#L8 "Error detection") for [a similar purpose](https://www.khronos.org/opengl/wiki/OpenGL_Error "OpenGL error handling")?

**Comment by EvilPudding** [comment]
Why? Every time `glCreateProgram` fails, `glerr` catches it.

**Comment by elfring** [comment]
Can it be nicer to call a function like “_check_gl_error” only after the detection of a failed function call?

**Comment by EvilPudding** [comment]
No, since there may be other errors even if `glCreateProgram` returns a non-zero value.

**Comment by elfring** [comment]
I guess that this should not be possible if the success (or failure) would be immediately checked for every function call (with a non-void return type).

**Comment by EvilPudding** [comment]
`glerr` checks for more than one thing, and it's positioned after every OpenGL call, I don't want to have it placed different depending on the call, most OpenGL functions don't even return anything.

Also, even if I were to do what you ask, the result wouldn't even be better. Engines take a lot of work, and there is a lot to implement and improve that actually matters in terms of performance and usability.

If I already have an error check, why would I place it inside another error check? What can I possibly gain from that?

**Comment by elfring** [comment]
>…, most OpenGL functions don't even return anything.

The check granularity will become more interesting then, won't it?


>Also, even if I were to do what you ask, the result wouldn't even be better.

How do you think about to call a function like “_check_gl_error” only in the failure case?


>What can I possibly gain from that?

Would you like to avoid the passing of extra information for better error reporting if it can be determined by the return value that a specific function call succeeded?

**Comment by EvilPudding** [comment]
I understand adding checks where there aren't any, and even then, if your computer is returning errors in `malloc`, then there's more problems with the computer than the engine.

As to adding checks on top of checks and trying to optimize the debugging tools seems like a waste of time.

**Comment by elfring** [comment]
Do you care for software safety and efficiency at the mentioned source code places?

**Comment by EvilPudding** [comment]
Adding checks does not increase software efficiency, it's the other way around. And `glerr` checks will probably be disabled for release compilations.

Go run any game engine without having memory and swap, the program crashes, as it should. What safety is increased by adding a `malloc` check?

As to the *GL* functions, checking errors is only good for debugging, no one is going to steal your credit card number by not checking *GL* errors.

**Comment by elfring** [comment]
>What safety is increased by adding a `malloc` check?

[More null pointer checks are required](https://wiki.sei.cmu.edu/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers "Do not dereference null pointers.") to achieve defined software behaviour.


>As to the GL functions, checking errors is only good for debugging, …

* I disagree to such a view. - I would generally like correct data displays.
* I imagine that the software can become a bit more efficient at specific source code places when less function calls will be performed because more return values can be checked also for successful function executions.

**Comment by EvilPudding** [comment]
>I disagree to such a view. - I would generally like correct data displays.

Data displayed incorrectly would never have anything to do with an OpenGL error but more to do with what data is sent. OpenGL errors make it so nothing displays.

>I imagine that the software can become a bit more efficient at specific source code places when less function calls will be performed because more return values can be checked also for successful function executions.

You imagining that is the exact definition of *premature optimization*. Thanks for the input though, I appreciate you taking the time.

**Comment by elfring** [comment]
>OpenGL errors make it so nothing displays.

Can such a data processing be also questionable?


>You imagining that is the exact definition of _premature optimization_.

I disagree here again.

The checking of a return value should be faster than to call an extra function, shouldn't it?

**Comment by EvilPudding** [comment]
>Can such a data processing be also questionable?

It can, but not in a way that would trigger an OpenGL error.

>The checking of a return value should be faster than to call an extra function, shouldn't it?

Like I said, OpenGL state error checking will be disabled for releases.

**Comment by elfring** [comment]
>Like I said, OpenGL state error checking will be disabled for releases.

I can find this software build configuration also questionable.
Do you try to follow Vulkan habits?

**Comment by EvilPudding** [comment]
>I can find this software build configuration also questionable.

Like I said, those checks are for debugging, like you said, function calls get expensive, many people do the same with `assert` where in release mode it does nothing.

**Comment by elfring** [comment]
Do you care for [advices around secure programming](https://en.wikipedia.org/wiki/Defensive_programming "Description for defensive programming")?

**Comment by EvilPudding** [comment]
I am all for using `strncpy` instead of `strcpy` (and do where it makes sense) to prevent crashes or bugs, which has nothing to do with the parts of the code you mentioned. What you mentioned prevents nothing.

**Comment by elfring** [comment]
>What you mentioned prevents nothing.

Secure programming guidelines provide other useful information for further software development considerations, don't they?

**Comment by EvilPudding** [comment]
I agree. When I say "what you mentioned" I don't mean "secure programming guidelines" I mean the specific input about `glCreateProgram` and `malloc` and `strdup`. Guidelines are useful because they help prevent errors. The only computers where `malloc` would generate an error would be too weak to run a game engine anyway.

**Comment by elfring** [comment]
It seems that there is another target conflict to consider between your software development understanding and other secure coding recommendations, isn't it?

**Comment by EvilPudding** [comment]
I don't understand the question. What do you mean *target conflict*?

**Comment by elfring** [comment]
You present a coding style with implementation details which are eventually not recommended.

**Comment by EvilPudding** [comment]
I think I passed my point across and you passed yours. Like I said before, thanks for the suggestions and input, but I'm moving on.
